### DISCUSSION TITLE
[FEAT] Fast API Middleware in Routers

### DISCUSSION THREAD

**Discussion by R5dan** [fastapi/fastapi#13046] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI,APIRouter
from extension import RouterExtension, AppExtension

app = FastAPI()
router = APIRouter()

@app.get("/")
async def app_route():
    #Some route

@app.get("/")
async def router_route():
    #Some route

router.add_extension(RouterExtension())
app.include_router(router, "/router")

#### Extension
from fastapi import Extension

class CustomExtension(Extension):
      def before_request(self, req):
          pass

      def after_request(self, resp):
          pass

      # Other methods for extensions
```

### Description

I want to be able to create an extension for Fast API so I don't need to write the same stuff on multiple routes (like accessing a database or authenticating a user). This could then help to speed up certain areas of development by having a single extension for it

### Operating System

Other

### Operating System Details

Should be able to work on any OS

### FastAPI Version

0.115.0

### Pydantic Version

2.10.1

### Python Version

3.13

### Additional Context

This is a feature request

**Comment by yvallois** [comment, marked as answer]
I'm not sure I understand what you want. Your snipplet looks like middleware, you can find the documentation here https://fastapi.tiangolo.com/tutorial/middleware/

**Reply by R5dan** [reply]
Yes, that seems to be what I'm looking for, do you know of any good docs for how to make one? And the different methods I can use?

I have looked at:
- [FastAPI middleware](https://fastapi.tiangolo.com/tutorial/middleware/): I could only find stuff about `app.middleware("http")`
- [Fast API Advanced Middleware](https://fastapi.tiangolo.com/advanced/middleware):  talks about existing middleware from FastAPI 
- [Starlette Middleware](https://www.starlette.io/middleware/): talks about it in a lot more detail but still doesn't actually describe the methods. It also talks about what middleware is for most of it.

**Reply by yvallois** [reply]
I mainly use the documents you mentioned.
If you don't want to use the interface proposed in the first documentation then you can find the minimum requirements here https://www.starlette.io/middleware/#writing-pure-asgi-middleware. 
You can also find some examples in this list https://github.com/florimondmanca/awesome-asgi

**Reply by R5dan** [reply]
Its not that I don't want to use them, more so that they don't seem to have any API documentation of what functions I can use. Can I only use `__call__`? Is that how I receive all requests? What do I do if I want to edit a pydantic model? Is there any easier class I can use? I know these might seem obvious but I have never made an ASGI middleware for any of the websites I've made: never needed one and it doesn't seem obvious to me from the docs?

**Reply by R5dan** [reply]
Also how do I add middleware only for certain Routers? Can I do it the same way I would an app?

**Reply by yvallois** [reply]
I haven't used middleware much so I'm not sure of my answers.

> Can I only use __call__?

Yes, This is the entry point that starlette will use

> Is that how I receive all requests?

All requests will pass through each successive middleware before arriving at the router

> What do I do if I want to edit a pydantic model?

If you're talking about the pydantic models used for body/query/form validation, they're used afterwards in the process, so no, you won't be able to modify them here.

> I know these might seem obvious but I have never made an ASGI middleware for any of the websites I've made: never needed one and it doesn't seem obvious to me from the docs?

You can take a look at how the various existing middleware is built if the documentation isn't enough for you, but I don't think you'll find any more detailed documentation.

> Also how do I add middleware only for certain Routers? Can I do it the same way I would an app?

I think that middleware is designed to be used at application level and not at router level. You can use https://fastapi.tiangolo.com/how-to/custom-request-and-route/#custom-apiroute-class-in-a-router to customise a router.

**Reply by R5dan** [reply]
Ok, thanks. I also couldn't find a way to add middleware to a specific router which is annoying (for authentication I don't want to force someone to be signed in on all routes) so am changing my feature request to that. Thanks for all the help you have given me.

**Reply by Kludex** [reply]
FastAPI still doesn't support middleware in `Router`s. Starlette does, so FastAPI just needs to add the parameter and pass it.

**Reply by R5dan** [reply]
Should I open an issue for this? I am happy to make the pull request as well.

**Reply by Kludex** [reply]
I opened the PR, Sebastian wants to do something else before it, or something.

You can look at my closed PRs, you should be able to find it - I'm on my phone.

**Reply by yvallois** [reply]
For authentication purposes, you can use dependency injection at router level, for exemple with a simple basic auth:
```
from fastapi import FastAPI, APIRouter, Depends
from fastapi.security import HTTPBasic, HTTPBasicCredentials

security = HTTPBasic()
app = FastAPI()


def basic_auth(credentials: HTTPBasicCredentials = Depends(security)) -> bool:
    # implement your logic here
    ...


router = APIRouter(dependencies=[Depends(basic_auth)])
app.include_router(router)
```

**Reply by R5dan** [reply]
> I opened the PR, Sebastian wants to do something else before it, or something.

Do you mean #11010? So this is unplanned for now?


> For authentication purposes, you can use dependency injection at router level, for exemple with a simple basic auth:

I know there are other ways, I could make a custom router etc. This would just be easiest.

**Reply by Kludex** [reply]
I've closed the PR, but yeah, that's the one I meant.
